File: packages/lcyt-backend/src/middleware/cors.js
const CORS_HEADERS = 'Content-Type, Authorization, X-Admin-Key, X-API-Key';
const CORS_METHODS = 'GET, POST, PUT, DELETE, PATCH, OPTIONS';

/**
 * Dynamic CORS middleware.
 *
 * - POST /live and GET /health: permissive — any origin may call (API key is the real gate)
 * - /dsk/* and /images: permissive — authenticated by X-API-Key or Bearer JWT, not session domain
 * - /events with ?token=: permissive — authenticated by token, origin is not the security gate
 * - /keys routes: permissive CORS — user project CRUD uses Bearer JWT from the browser; admin key is protected by the route handler
 * - All other routes: look up sessions by domain; allow only registered origins
 *
 * @param {import('../store.js').SessionStore} store
 */
export function createCorsMiddleware(store) {
  return function corsMiddleware(req, res, next) {
    const origin = req.headers['origin'];
    const path = req.path;
    const method = req.method;

    // Free-tier key signup — any origin, POST only (must check before the /keys block)
    if (method === 'POST' && path === '/keys' && 'freetier' in req.query) {
      if (origin) {
        res.setHeader('Access-Control-Allow-Origin', origin);
        res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
        res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
        res.setHeader('Access-Control-Allow-Credentials', 'true');
      }
      return next();
    }

    // /keys routes — user project CRUD uses Bearer JWT (needs CORS for browser apps like
    // the Projects page at app.lcyt.fi). Admin operations are protected by X-Admin-Key in
    // the route handler; CORS being open here does not weaken that protection.
    if (path.startsWith('/keys')) {
      if (origin) {
        res.setHeader('Access-Control-Allow-Origin', origin);
        res.setHeader('Access-Control-Allow-Methods', CORS_METHODS);
        res.setHeader('Access-Control-Allow-Headers', CORS_HEADERS);
        res.setHeader('Access-Control-Allow-Credentials', 'true');
      }
      if (method === 'OPTIONS') {
        res.sendStatus(204);
        return;
      }
      return next();
    }

    // Permissive routes — any origin can call; the auth header/token is the real gate
    const isPermissive =
      (method === 'POST' && path === '/live') ||
      (method === 'GET' && path === '/health') ||
      (method === 'GET' && path === '/contact') ||
      // DSK routes: authenticated by X-API-Key or Bearer JWT, not by session domain
      path.startsWith('/dsk') ||
      path.startsWith('/images') ||
      // /events with ?token= is token-authenticated; origin is not the security gate
      (path === '/events' && req.query.token) ||
      method === 'OPTIONS';

    if (isPermissive) {
      if (origin) {
        res.setHeader('Access-Control-Allow-Origin', origin);
        res.setHeader('Access-Control-Allow-Methods', CORS_METHODS);
        res.setHeader('Access-Control-Allow-Headers', CORS_HEADERS);
        res.setHeader('Access-Control-Allow-Credentials', 'true');
      }
      if (method === 'OPTIONS') {
        res.sendStatus(204);
        return;
      }
      return next();
    }

    // Dynamic origin matching for session-authenticated routes
    if (origin) {
      const matches = store.getByDomain(origin);
      if (matches.length > 0) {
        res.setHeader('Access-Control-Allow-Origin', origin);
        res.setHeader('Access-Control-Allow-Methods', CORS_METHODS);
        res.setHeader('Access-Control-Allow-Headers', CORS_HEADERS);
        res.setHeader('Access-Control-Allow-Credentials', 'true');
      }
      // No match → omit CORS headers (browser will block the request)
    }

    if (method === 'OPTIONS') {
      res.sendStatus(204);
      return;
    }

    next();
  };
}

File: python-packages/lcyt-backend/lcyt_backend/middleware/cors.py
"""Dynamic CORS middleware for lcyt-backend."""

from flask import Flask, request

from ..store import SessionStore

_ALLOWED_HEADERS = "Content-Type, Authorization, X-Admin-Key"
_ALLOWED_METHODS = "GET, POST, DELETE, PATCH, OPTIONS"


def register_cors_middleware(app: Flask, store: SessionStore) -> None:
    """Attach dynamic CORS logic to the Flask app via after_request hooks.

    Rules (mirrors the Node.js backend):
      - /keys routes          — no CORS headers (server-side admin only)
      - POST /live, GET /health, OPTIONS — permissive (any origin)
      - All other routes      — dynamic: allow origin only if it has an active session

    Args:
        app: Flask application instance.
        store: Active session store used to look up registered origins.
    """

    @app.before_request
    def handle_preflight():
        """Short-circuit OPTIONS preflight requests."""
        if request.method != "OPTIONS":
            return None

        origin = request.headers.get("Origin", "")
        path = request.path

        # Admin endpoints — no CORS even for OPTIONS
        if path.startswith("/keys"):
            return ("", 204)

        if origin:
            response_headers = {
                "Access-Control-Allow-Methods": _ALLOWED_METHODS,
                "Access-Control-Allow-Headers": _ALLOWED_HEADERS,
                "Access-Control-Allow-Credentials": "true",
            }

            # Permissive routes or a registered origin
            if _is_permissive_route(request.method, path) or store.get_by_domain(origin):
                response_headers["Access-Control-Allow-Origin"] = origin

            return ("", 204, response_headers)

        return ("", 204)

    @app.after_request
    def add_cors_headers(response):
        """Add CORS headers to non-preflight responses."""
        origin = request.headers.get("Origin", "")
        path = request.path
        method = request.method

        # Admin endpoints — no CORS headers
        if path.startswith("/keys"):
            return response

        if not origin:
            return response

        if _is_permissive_route(method, path):
            response.headers["Access-Control-Allow-Origin"] = origin
            response.headers["Access-Control-Allow-Methods"] = _ALLOWED_METHODS
            response.headers["Access-Control-Allow-Headers"] = _ALLOWED_HEADERS
            response.headers["Access-Control-Allow-Credentials"] = "true"
        elif store.get_by_domain(origin):
            response.headers["Access-Control-Allow-Origin"] = origin
            response.headers["Access-Control-Allow-Methods"] = _ALLOWED_METHODS
            response.headers["Access-Control-Allow-Headers"] = "Content-Type, Authorization"
            response.headers["Access-Control-Allow-Credentials"] = "true"
        # else: no match → omit CORS headers (browser will block the request)

        return response


def _is_permissive_route(method: str, path: str) -> bool:
    return (method == "POST" and path == "/live") or (method == "GET" and path == "/health")
